astroquery/mast/core.py

```diff
@@ -63,59 +77,258 @@ def _prepare_service_request_string(json_obj):
 
         super(MastClass, self).__init__()
 
         self._MAST_REQUEST_URL = conf.server + "/api/v0/invoke"
         self._MAST_DOWNLOAD_URL = conf.server + "/api/v0/download/file/"
         self._COLUMNS_CONFIG_URL = conf.server + "/portal/Mashup/Mashup.asmx/columnsconfig"
 
+        # shibbolith urls
+        self._SP_TARGET = conf.server + "/api/v0/Mashup/Login/login.html"
+        self._IDP_ENDPOINT = conf.ssoserver + "/idp/profile/SAML2/SOAP/ECP"
+        self._SESSION_INFO_URL = conf.server + "/Shibboleth.sso/Session"
+
         self.TIMEOUT = conf.timeout
         self.PAGESIZE = conf.pagesize
 
+        self._column_configs = {}
+        self._current_service = None
+
+        if username or session_token:
+            self.login(username, password, session_token)
+
+    def _attach_cookie(self, session_token):  # pragma: no cover
+        """
+        Attaches a valid shibboleth session cookie to the current session.
+
+        Parameters
+        ----------
+        session_token : dict or  `http.cookiejar.Cookie`
+            A valid MAST shibboleth session cookie.
+        """
+
+        # clear any previous shib cookies
+        self._session.cookies.clear_session_cookies()
+
+        if isinstance(session_token, Cookie):
+            # check it's a shibsession cookie
+            if "shibsession" not in session_token.name:
+                raise LoginError("Invalid session token")
+
+            # add cookie to session
+            self._session.cookies.set_cookie(session_token)
+
+        elif isinstance(session_token, dict):
+            if len(session_token) > 1:
+                warnings.warn("Too many entries in token dictionary, only shibsession cookie will be used",
+                              InputWarning)
+
+            # get the shibsession cookie
+            value = None
+            for name in session_token.keys():
+                if "shibsession" in name:
+                    value = session_token[name]
+                    break
+
+            if not value:
+                raise LoginError("Invalid session token")
+
+            # add cookie to session
+            self._session.cookies.set(name, value)
+
+        else:
+            # raise datatype error
+            raise LoginError("Session token must be given as a dictionary or http.cookiejar.Cookie object")
+
+        # Print session info
+        # get user information
+        response = self._session.request("GET", self._SESSION_INFO_URL)
+
+        if response.status_code != 200:
+            print("Status code:", response.status_code)
+            print("Authentication failed!")
+            return False
+
+        exp = re.findall(r'<strong>Session Expiration \(barring inactivity\):</strong> (.*?)\n', response.text)
+        if len(exp) == 0:
+            print(response.text)
+            print("Authentication failed!")
+            return False
+        else:
+            exp = exp[0]
+
+        print("Authentication successful!")
```

Comment:
is there a way to silence the prints? Either you could use a "verbose" keyword or the logging machinery? These can be then be simple ``log.info``s and the authentication failed ones ``log.warning``s?

Reply:
Done.